File: 0132-palindrome-partitioning-ii/0132-palindrome-partitioning-ii.py
class Solution:
    def minCut(self, s: str) -> int:
        
        n = len(s)
        
        dp = [[False] * n for _ in range(n)]
        
        for i in range(n):
            dp[i][i] = True
            
        for i in range(n-1):
            if s[i] == s[i+1]:
                dp[i][i+1] = True
                
        for length in range(3, n+1):
            for i in range(n-length+1):
                j = i + length - 1
                if s[i] == s[j] and dp[i+1][j-1]:
                    dp[i][j] = True
                    
        cuts = [inf] * n
        cuts[0] = 0
        
        for i in range(1, n):
            if dp[0][i]:
                cuts[i] = 0
            else:
                for j in range(i):
                    if dp[j+1][i]:
                        cuts[i] = min(cuts[i], 1 + cuts[j])
        
        return cuts[n-1]

File: 0132-palindrome-partitioning-ii/0132-palindrome-partitioning-ii.js
/**
 * @param {string} s
 * @return {number}
 */
var minCut = function(s) {
    const n = s.length
    const dp = Array(n).fill(0).map(() => Array(n).fill(false))
    for (let i=0; i < n; i++) {
        dp[i][i] = true
    }
    for (let i=0; i < n-1; i++) {
        if (s[i] === s[i+1]) {
            dp[i][i+1] = true
        }
    }
    for (let len=3; len <= n; len++) {
        for (let i=0; i < n - len + 1; i++) {
            const j = i + len - 1
            if (s[i] === s[j] && dp[i+1][j-1]) {
                dp[i][j] = true
            }
        }
    }
    const cuts = Array(n).fill(Infinity)
    cuts[0] = 0
    for (let i=1; i < n; i++) {
        if (dp[0][i]) {
            cuts[i] = 0
        } else {
            for (let j=0; j < i; j++) {
                if (dp[j+1][i]) {
                    cuts[i] = Math.min(cuts[i], 1 + cuts[j])
                }
            }
        }
    }
    return cuts[n-1]
};